index.bs

```diff
@@ -4091,6 +4104,138 @@ The following extended attributes are applicable to [=iterable declarations=]:
+                  :: the username in <b>this</b>'s open sessions that is next in lexicographical
+                     order, after <b>current state</b>
+            </dl>
+        1.  If there is no such |key|, then:
+            1.  Resolve |promise| with undefined.
+        1.  Otherwise:
+            1.  Let |session| be the <code class="idl">Session</code> object corresponding to |key|.
+            1.  Resolve |promise| with (|username|, |session|, |username|).
+        1.  Return |promise|.
+
+    </blockquote>
+
+    In the ECMAScript language binding, the [=interface prototype object=] for the
+    <code class=idl>SessionManager</code> [=interface=] has a <code>values</code> method that is a
+    function, which, when invoked, returns an asynchronous iterator object that itself has a
+    <code>next</code> method that returns the next value to be iterated over.
+    It has <code>keys</code> and <code>entries</code> methods that iterate over the usernames of
+    session objects and (username, <code class=idl>Session</code>) object pairs, respectively.
+    It also has a {{@@asyncIterator}} method that allows a <code class="idl">SessionManager</code>
+    to be used in a <code>for await..of</code> loop that has the same value as the
+    <code>entries</code> method:
+
+    <pre highlight="js">
+        // Get an instance of SessionManager.
+        // Assume that it has sessions for two users, "anna" and "brian".
+        var sm = getSessionManager();
+
+        typeof SessionManager.prototype.values;            // Evaluates to "function"
+        var it = sm.values();                              // values() returns an iterator object
+        typeof it.next;                                    // Evaluates to "function"
+
+        // This loop will log "anna" and then "brian".
+        for await (let username of sm.keys()) {
+          console.log(username);
+        }
+
+        // Yet another way of accomplishing the same.
+        for await (let [username, session] of sm) {
+          console.log(username);
+        }
+    </pre>
+</div>
+
+An [=interface=] must not have more than one [=asynchronously iterable declaration=].
+The [=inherited interfaces=] of an [=interface=] with an [=asynchronously iterable declaration=]
+must not also have an [=asynchronously iterable declaration=].
+An [=interface=] with an [=asynchronously iterable declaration=] and its [=inherited interfaces=]
+must not have a [=maplike declaration=], [=setlike declaration=], or [=iterable declaration=].
+
+The following extended attributes are applicable to [=asynchronously iterable declarations=]:
+[{{Exposed}}],
+[{{SecureContext}}].
+
+Issue: these [=extended attributes=] are not currently taken into account.
+When they are, the effect will be as you would expect.
+
+<pre class="grammar" id="prod-AsyncIterable">
+    AsyncIterable :
+        "async_iterable" "&lt;" TypeWithExtendedAttributes "," TypeWithExtendedAttributes "&gt;" ";"
```

Comment:
For https://wicg.github.io/native-file-system/#filesystemdirectoryhandle I believe we'll want a version that only does values.

For https://streams.spec.whatwg.org/#rs-asynciterator / https://streams.spec.whatwg.org/#rs-get-iterator we'll want a version that only does values and also can be aliased to a function that takes arguments.

For the values-only version I'd stay conservative and not add keys/values/entries for now as this is uncharted territory. (Even though sync iterables do that even for values-only cases like `Set`.)

All this can be done as a follow-up, but it may be helpful to keep it in mind while designing.

Reply:
Streams will also need some side effects on acquiring the iterator (i.e. calling the function)

Reply:
Added the side effects, but I'd like to defer full Streams support to a followup.